README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# intradayModel

Our package uses state-of-the-art state-space models to facilitate the
modeling and forecasting of financial intraday signals. It currently
offers a univariate model for intraday trading volume, with new features
on intraday volatility and multivariate models in development. It is a
valuable tool for anyone interested in exploring intraday, algorithmic,
and high-frequency trading.

## Installation

The package can be installed from
[GitHub](https://github.com/convexfi/intradayModel):

``` r
# install development version from GitHub
devtools::install_github("convexfi/intradayModel")
```

Please cite **intradayModel** in publications:

``` r
citation("intradayModel")
```

## Quick Start

To get started, we load our package and sample data: the 15-minute
intraday trading volume of AAPL from 2019-01-02 to 2019-06-28, covering
124 trading days. We use the first 104 trading days for fitting, and the
last 20 days for evaluation of forecasting performance.

``` r
library(intradayModel)
data(volume_aapl)
volume_aapl[1:5, 1:5] # print the head of data
#>          2019-01-02 2019-01-03 2019-01-04 2019-01-07 2019-01-08
#> 09:30 AM   10142172    3434769   20852127   15463747   14719388
#> 09:45 AM    5691840   19751251   13374784    9962816    9515796
#> 10:00 AM    6240374   14743180   11478596    7453044    6145623
#> 10:15 AM    5273488   14841012   16024512    7270399    6031988
#> 10:30 AM    4587159   18041115    8686059    7130980    5479852

volume_aapl_training <- volume_aapl[, 1:104]
volume_aapl_testing <- volume_aapl[, 105:124]
```

Next, we fit a univariate state-space model using `fit_volume()`
function.

``` r
model_fit <- fit_volume(volume_aapl_training)
```

Once the model is fitted, we can analyze the hidden components of any
intraday volume based on all its observations. By calling
`decompose_volume()` function with `purpose = "analysis"`, we obtain the
smoothed daily, seasonal, and intraday dynamic components. It involves
incorporating both past and future observations to refine the state
estimates.

``` r
analysis_result <- decompose_volume(purpose = "analysis", model_fit, volume_aapl_training)

# visualization
plots <- generate_plots(analysis_result)
plots$log_components
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" style="display: block; margin: auto;" />

To see how well our model performs on new data, we call
`forecast_volume()` function to do one-bin-ahead forecast on the testing
set.

``` r
forecast_result <- forecast_volume(model_fit, volume_aapl_testing)

# visualization
plots <- generate_plots(forecast_result)
plots$original_and_forecast
```

<img src="man/figures/README-unnamed-chunk-7-1.png" width="100%" style="display: block; margin: auto;" />

## Contributing

We welcome all sorts of contributions. Please feel free to open an issue
to report a bug or discuss a feature request.

## Citation

If you make use of this software please consider citing:

-   Chen, R., Feng, Y., and Palomar, D. (2016). Forecasting intraday
    trading volume: A Kalman filter approach.
    <https://dx.doi.org/10.2139/ssrn.3101695>

## Links

Package: [GitHub](https://github.com/convexfi/intradayModel)

Vignette:
[GitHub-vignette](https://htmlpreview.github.io/?https://github.com/convexfi/intradayModel/blob/master/vignettes/intradayModel.html).


NEWS.md
## Changes in intradayModel version 0.0.1 (2023-05-19)

* Initial bare-bone implementation for intraday volume (Chen 2016).


R/plotting.R
#' @title Plot Analysis and Forecast Result
#'
#' @description Autoplot the results of analysis and forecast.
#'
#' @param analysis_forecast_result Analysis/forecast result from \code{use_model}.
#'
#' @return A list of \code{patchwork} objects:
#'        \itemize{
#'        \item{\code{log_components}: }{Plot of components of intraday signal in their log10 scale;}
#'        \item{\code{original_and_resultant}: }{Plot of the original and the resulant signals.}}
#'
#' @import patchwork
#' @import ggplot2
#' @import scales
#' @importFrom magrittr %>%
#' @examples
#' \dontrun{
#'
#' data(aapl_volume)
#' aapl_volume_training <- aapl_volume[, 1:104]
#' aapl_volume_testing <- aapl_volume[, 105:124]
#'
#' # Obtain smoothing and forecasting result
#' model_fit <- fit_volume(aapl_volume_training)
#' analysis_result <- use_model(purpose = "analysis", model_fit, aapl_volume_training)
#' forecast_result <- use_model(purpose = "forecast", model_fit, aapl_volume_testing)
#' 
#' # Plot components
#' autoplot(analysis_result)
#' autoplot(forecast_result)
#' }
#' @export
autoplot <- function(analysis_forecast_result) {
  plot_list <- list()
  
  plot_list$log_components <- plot_components(analysis_forecast_result)
  plot_list$original_and_resultant <- plot_performance(analysis_forecast_result)
  
  return(plot_list)
}

plot_components <- function(analysis_forecast_result) {
  i <- original <- daily <- seasonal <- dynamic <- NULL
  components <- analysis_forecast_result[["components"]]

  plt_data <-
    data.frame(
      original = analysis_forecast_result$original_signal,
      daily = components[[grep("daily", names(components))]],
      seasonal = components[[grep("seasonal", names(components))]],
      dynamic = components[[grep("dynamic", names(components))]],
      residual = components$residual
    )
  plt_data_log <- log10(plt_data)
  plt_data_log$i <- plt_data$i <- c(1:nrow(plt_data))

  text_size <- 10
  p1 <- plt_data_log %>%
    ggplot() +
    geom_line(aes(x = i, y = original), alpha = 0.8, color = "steelblue", size = 0.4) +
    ylab("Original") +
    theme_bw() +
    theme(
      axis.title = element_text(size = text_size, face = "bold"),
      legend.position = "right",
      legend.justification = c(0, 1),
      legend.box.just = "left",
      legend.margin = margin(8, 8, 8, 8),
      legend.text = element_text(size = text_size, face = "bold"),
      legend.key.size = unit(1, "cm"),
      plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
      axis.title.x = element_blank(),
      axis.text.x = element_blank()
    )

  p2 <- plt_data_log %>%
    ggplot() +
    geom_line(aes(x = i, y = daily), alpha = 0.8, color = "steelblue", size = 0.6) +
    ylab("Daily") +
    theme_bw() +
    theme(
      axis.title = element_text(size = text_size, face = "bold"),
      legend.position = "right",
      legend.justification = c(0, 1),
      legend.box.just = "left",
      legend.margin = margin(8, 8, 8, 8),
      legend.text = element_text(size = text_size, face = "bold"),
      legend.key.size = unit(1, "cm"),
      plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
      axis.title.x = element_blank(),
      axis.text.x = element_blank()
    )

  p3 <- plt_data_log %>%
    ggplot() +
    geom_line(aes(x = i, y = seasonal), alpha = 0.8, color = "steelblue", size = 0.4) +
    ylab("Seasonal") +
    theme_bw() +
    theme(
      axis.title = element_text(size = text_size, face = "bold"),
      legend.position = "right",
      legend.justification = c(0, 1),
      legend.box.just = "left",
      legend.margin = margin(8, 8, 8, 8),
      legend.text = element_text(size = text_size, face = "bold"),
      legend.key.size = unit(1, "cm"),
      plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
      axis.title.x = element_blank(),
      axis.text.x = element_blank()
    )

  p4 <- plt_data_log %>%
    ggplot() +
    geom_line(aes(x = i, y = dynamic), alpha = 0.8, color = "steelblue", size = 0.4) +
    ylab("Intraday\nDynamic") +
    theme_bw() +
    theme(
      axis.title = element_text(size = text_size, face = "bold"),
      legend.position = "right",
      legend.justification = c(0, 1),
      legend.box.just = "left",
      legend.margin = margin(8, 8, 8, 8),
      legend.text = element_text(size = text_size, face = "bold"),
      legend.key.size = unit(1, "cm"),
      plot.title = element_text(size = 18, face = "bold", hjust = 0.5),
      axis.title.x = element_blank(),
      axis.text.x = element_blank()
    )

  p5 <- plt_data_log %>%
    ggplot() +
    geom_line(aes(x = i, y = residual), alpha = 0.8, color = "steelblue", size = 0.4) +
    ylab("Residual") +
    theme_bw() +
    xlab("time (bins)") +
    theme(
      axis.title.x = element_text(size = 12, face = "bold"),
      axis.title.y = element_text(size = text_size, face = "bold"),
      legend.position = "right",
      legend.justification = c(0, 1),
      legend.box.just = "left",
      legend.margin = margin(8, 8, 8, 8),
      legend.text = element_text(size = text_size, face = "bold"),
      legend.key.size = unit(1, "cm"),
      plot.title = element_text(size = 18, face = "bold", hjust = 0.5)
    )
  
  p <- p1 / p2 / p3 / p4 / p5 +
    plot_annotation(
      title = "Decomposition of intraday signal",
      theme = theme(plot.title = element_text(size = 16, face = "bold", hjust = 0.5))
    )

  return(p)
}

plot_performance <- function(analysis_forecast_result) {
  i <- value <- variable <- NULL

  # determine type
  if (sum(grepl("smooth", names(analysis_forecast_result)))) {
    type <- "smooth"
    title <- "Analysis result"
    legend_name <- "model smooth fit"
  } else {
    type <- "forecast"
    title <- "One-bin-ahead forecast"
    legend_name <- "forecast"
  }

  plt_data <-
    data.frame(
      original = analysis_forecast_result$original_signal,
      output = analysis_forecast_result[[grep(type, names(analysis_forecast_result))]]
    )
#  plt_data_log <- log(plt_data)
#  plt_data_log$i <- plt_data$i <- c(1:nrow(plt_data))

  plt_data$i <- c(1:nrow(plt_data))
  
  plt_reshape <- plt_data %>%
    reshape2::melt(
      id.vars = c("i"),
      variable.name = "variable", value.name = "value"
    )

  text_size <- 14
  p <- plt_reshape %>%
    ggplot() +
    geom_line(aes(x = i, y = value, color = variable), alpha = 0.8, size = 0.4) +
    scale_colour_manual(values = c(original = "steelblue", output = "#FD6467"), labels = c("original", legend_name)) +
    scale_y_log10(
      breaks = trans_breaks("log10", function(x) 10^x),
      labels = trans_format("log10", math_format(10^.x))
    ) +
    xlab("time (bins)") +
    ylab("Intraday Signal") +
    theme_bw() +
    theme(
      axis.title = element_text(size = text_size, face = "bold"),
      legend.position = "bottom",
      legend.text = element_text(size = text_size, face = "bold"),
      legend.title = element_blank(),
      plot.title = element_text(size = 16, face = "bold", hjust = 0.5)
    ) +
    plot_annotation(
      title = title,
      theme = theme(plot.title = element_text(size = 16, face = "bold", hjust = 0.5))
    )

  return(p)
}
